In [29]:
import pandas as pd

In [30]:
df = pd.read_csv("breast_drug_cancer.csv")

In [31]:
df

DRUG_ID  COSMIC_ID        CCLE_Name  TSPAN6 (7105)  TNMD (64102)  \
0        1003     909907    ZR7530_BREAST       3.472488           0.0   
1        1004     909907    ZR7530_BREAST       3.472488           0.0   
2        1005     909907    ZR7530_BREAST       3.472488           0.0   
3        1006     909907    ZR7530_BREAST       3.472488           0.0   
4        1007     909907    ZR7530_BREAST       3.472488           0.0   
...       ...        ...              ...            ...           ...   
4323     2045     908121  MDAMB361_BREAST       1.855990           0.0   
4324     2046     908121  MDAMB361_BREAST       1.855990           0.0   
4325     2048     908121  MDAMB361_BREAST       1.855990           0.0   
4326     2106     908121  MDAMB361_BREAST       1.855990           0.0   
4327     2169     908121  MDAMB361_BREAST       1.855990           0.0   

      DPM1 (8813)  SCYL3 (57147)  C1orf112 (55732)  FGR (2268)  CFH (3075)  \
0        5.959306       3.878725          3.646163    0.000000    0.042644   
1        5.959306       3.878725          3.646163    0.000000    0.042644   
2        5.959306       3.878725          3.646163    0.000000    0.042644   
3        5.959306       3.878725          3.646163    0.000000    0.042644   
4        5.959306       3.878725          3.646163    0.000000    0.042644   
...           ...            ...               ...         ...         ...   
4323     6.736740       2.885574          3.766595    0.028569    0.879706   
4324     6.736740       2.885574          3.766595    0.028569    0.879706   
4325     6.736740       2.885574          3.766595    0.028569    0.879706   
4326     6.736740       2.885574          3.766595    0.028569    0.879706   
4327     6.736740       2.885574          3.766595    0.028569    0.879706   

      ...  246  247  248  249  250  251  252  253  254  255  
0     ...    0    0    1    0    1    0    0    0    0    0  
1     ...    0    1    0    1    0    1    0    0    0    0  
2     ...    0    0    0    0    0    0    0    0    0    0  
3     ...    0    0    0    0    1    1    0    0    0    0  
4     ...    0    0    1    1    1    1    0    0    0    0  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
4323  ...    0    0    0    0    1    0    0    0    0    0  
4324  ...    0    0    0    0    0    1    0    0    0    1  
4325  ...    0    1    0    1    1    1    0    0    0    0  
4326  ...    0    0    0    1    0    0    0    0    0    1  
4327  ...    1    0    0    0    0    0    0    0    0    0  

[4328 rows x 19484 columns]

In [33]:
to_drop = ['CCLE_Name', 'DRUG_NAME', 'isosmiles','PubCHEM']

In [34]:
#Remove drug id column's data
df.drop(to_drop, inplace=True, axis=1)

In [36]:
df

DRUG_ID  COSMIC_ID  TSPAN6 (7105)  TNMD (64102)  DPM1 (8813)  \
0        1003     909907       3.472488           0.0     5.959306   
1        1004     909907       3.472488           0.0     5.959306   
2        1005     909907       3.472488           0.0     5.959306   
3        1006     909907       3.472488           0.0     5.959306   
4        1007     909907       3.472488           0.0     5.959306   
...       ...        ...            ...           ...          ...   
4323     2045     908121       1.855990           0.0     6.736740   
4324     2046     908121       1.855990           0.0     6.736740   
4325     2048     908121       1.855990           0.0     6.736740   
4326     2106     908121       1.855990           0.0     6.736740   
4327     2169     908121       1.855990           0.0     6.736740   

      SCYL3 (57147)  C1orf112 (55732)  FGR (2268)  CFH (3075)  FUCA2 (2519)  \
0          3.878725          3.646163    0.000000    0.042644      4.711495   
1          3.878725          3.646163    0.000000    0.042644      4.711495   
2          3.878725          3.646163    0.000000    0.042644      4.711495   
3          3.878725          3.646163    0.000000    0.042644      4.711495   
4          3.878725          3.646163    0.000000    0.042644      4.711495   
...             ...               ...         ...         ...           ...   
4323       2.885574          3.766595    0.028569    0.879706      5.773733   
4324       2.885574          3.766595    0.028569    0.879706      5.773733   
4325       2.885574          3.766595    0.028569    0.879706      5.773733   
4326       2.885574          3.766595    0.028569    0.879706      5.773733   
4327       2.885574          3.766595    0.028569    0.879706      5.773733   

      ...  246  247  248  249  250  251  252  253  254  255  
0     ...    0    0    1    0    1    0    0    0    0    0  
1     ...    0    1    0    1    0    1    0    0    0    0  
2     ...    0    0    0    0    0    0    0    0    0    0  
3     ...    0    0    0    0    1    1    0    0    0    0  
4     ...    0    0    1    1    1    1    0    0    0    0  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
4323  ...    0    0    0    0    1    0    0    0    0    0  
4324  ...    0    0    0    0    0    1    0    0    0    1  
4325  ...    0    1    0    1    1    1    0    0    0    0  
4326  ...    0    0    0    1    0    0    0    0    0    1  
4327  ...    1    0    0    0    0    0    0    0    0    0  

[4328 rows x 19480 columns]

In [62]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.metrics import mean_squared_error, mean_absolute_error
from scipy.stats import pearsonr
from sklearn.metrics import r2_score

In [63]:
# Split data into features (X) and target variable (y)
X = df.drop(columns=['DRUG_ID', 'COSMIC_ID', 'LN_IC50'])
y = df['LN_IC50']

In [64]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [65]:
# Standardize features by removing the mean and scaling to unit variance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [69]:
# Hyperparameter tuning
# Define configurations for neurons and activation functions for each layer
neurons_list_layer1 = [64, 128, 256]
activation_list_layer1 = ['relu', 'tanh', 'sigmoid']
neurons_list_layer2 = [32, 64, 128]
activation_list_layer2 = ['relu', 'tanh', 'sigmoid']

In [70]:
# Initialize variables to store best results
best_rmse = float('inf')
best_config = None

In [71]:
# Loop over configurations for layer 1
for neurons_layer1 in neurons_list_layer1:
    for activation_layer1 in activation_list_layer1:
        # Loop over configurations for layer 2
        for neurons_layer2 in neurons_list_layer2:
            for activation_layer2 in activation_list_layer2:
                print(f"Training model with {neurons_layer1} neurons and {activation_layer1} activation function for layer 1...")
                print(f"                   and {neurons_layer2} neurons and {activation_layer2} activation function for layer 2...")
                
                # Build the deep learning model
                model = Sequential()
                model.add(Dense(neurons_layer1, activation=activation_layer1, input_shape=(X_train.shape[1],)))
                model.add(Dense(neurons_layer2, activation=activation_layer2))
                model.add(Dense(1))

                # Compile the model
                model.compile(optimizer='adam', loss='mean_squared_error')

                # Early stopping
                early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

                # Train the model with early stopping
                model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_split=0.2, callbacks=[early_stopping], verbose=0)

                # Predict on the test set
                y_pred = model.predict(X_test_scaled)
                
                # Convert predictions to numpy arrays
                y_test_np = np.array(y_test)
                y_pred_np = np.squeeze(y_pred)

                # Remove NaN values from predictions
                nan_indices = np.isnan(y_pred_np)
                y_test_np = y_test_np[~nan_indices]
                y_pred_np = y_pred_np[~nan_indices]

                # Evaluation metrics
                rmse = np.sqrt(mean_squared_error(y_test_np, y_pred_np))
                
                # Print RMSE for current configuration
                print(f"RMSE for {neurons_layer1} neurons and {activation_layer1} activation function for layer 1,")
                print(f"            and {neurons_layer2} neurons and {activation_layer2} activation function for layer 2: {rmse}")

                # Check if current configuration is the best
                if rmse < best_rmse:
                    best_rmse = rmse
                    best_config = (neurons_layer1, activation_layer1, neurons_layer2, activation_layer2)

print("Best configuration found:", best_config)
print("Best RMSE:", best_rmse)

Training model with 64 neurons and relu activation function for layer 1...
                   and 32 neurons and relu activation function for layer 2...
RMSE for 64 neurons and relu activation function for layer 1,
            and 32 neurons and relu activation function for layer 2: 1.5782776862003391
Training model with 64 neurons and relu activation function for layer 1...
                   and 32 neurons and tanh activation function for layer 2...
RMSE for 64 neurons and relu activation function for layer 1,
            and 32 neurons and tanh activation function for layer 2: 2.7489429915300647
Training model with 64 neurons and relu activation function for layer 1...
                   and 32 neurons and sigmoid activation function for layer 2...
RMSE for 64 neurons and relu activation function for layer 1,
            and 32 neurons and sigmoid activation function for layer 2: 2.7204427019538056
Training model with 64 neurons and relu activation function for layer 1...
          

RMSE for 128 neurons and relu activation function for layer 1,
            and 32 neurons and relu activation function for layer 2: 1.508501242791765
Training model with 128 neurons and relu activation function for layer 1...
                   and 32 neurons and tanh activation function for layer 2...
RMSE for 128 neurons and relu activation function for layer 1,
            and 32 neurons and tanh activation function for layer 2: 2.762908779243825
Training model with 128 neurons and relu activation function for layer 1...
                   and 32 neurons and sigmoid activation function for layer 2...
RMSE for 128 neurons and relu activation function for layer 1,
            and 32 neurons and sigmoid activation function for layer 2: 2.743950133214977
Training model with 128 neurons and relu activation function for layer 1...
                   and 64 neurons and relu activation function for layer 2...
RMSE for 128 neurons and relu activation function for layer 1,
            and 64 

RMSE for 256 neurons and relu activation function for layer 1,
            and 32 neurons and relu activation function for layer 2: 1.712637978508939
Training model with 256 neurons and relu activation function for layer 1...
                   and 32 neurons and tanh activation function for layer 2...
RMSE for 256 neurons and relu activation function for layer 1,
            and 32 neurons and tanh activation function for layer 2: 2.7573131127128803
Training model with 256 neurons and relu activation function for layer 1...
                   and 32 neurons and sigmoid activation function for layer 2...
RMSE for 256 neurons and relu activation function for layer 1,
            and 32 neurons and sigmoid activation function for layer 2: 2.7659921341374583
Training model with 256 neurons and relu activation function for layer 1...
                   and 64 neurons and relu activation function for layer 2...
RMSE for 256 neurons and relu activation function for layer 1,
            and 6

In [72]:
# Build the deep learning model using the best parameters
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dense(128, activation='relu'))
model.add(Dense(1))

In [73]:
# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

In [77]:
# Early stopping
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

In [78]:
# Train the model with early stopping
history = model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_split=0.2, callbacks=[early_stopping], verbose=1)

Epoch 1/100
87/87 [==============================] - 2s 18ms/step - loss: 1.4706 - val_loss: 1.9162
Epoch 2/100
87/87 [==============================] - 1s 13ms/step - loss: 1.4053 - val_loss: 1.9930
Epoch 3/100
87/87 [==============================] - 1s 13ms/step - loss: 1.4488 - val_loss: 1.9762
Epoch 4/100
87/87 [==============================] - 1s 12ms/step - loss: 1.5021 - val_loss: 2.0640
Epoch 5/100
87/87 [==============================] - 1s 14ms/step - loss: 1.4077 - val_loss: 1.8767
Epoch 6/100
87/87 [==============================] - 1s 13ms/step - loss: 1.3327 - val_loss: 2.0573
Epoch 7/100
87/87 [==============================] - 1s 13ms/step - loss: 1.3380 - val_loss: 2.0642
Epoch 8/100
87/87 [==============================] - 1s 13ms/step - loss: 1.4284 - val_loss: 2.0299
Epoch 9/100
87/87 [==============================] - 1s 13ms/step - loss: 1.3282 - val_loss: 1.9664
Epoch 10/100
87/87 [==============================] - 1s 13ms/step - loss: 1.2795 - val_loss: 1.9687

In [79]:
# Predict on the test set
y_pred = model.predict(X_test_scaled)

In [80]:
# Convert predictions to numpy arrays
y_test_np = np.array(y_test)
y_pred_np = np.squeeze(y_pred)

# Remove NaN values from predictions
nan_indices = np.isnan(y_pred_np)
y_test_np = y_test_np[~nan_indices]
y_pred_np = y_pred_np[~nan_indices]

# Evaluation metrics
rmse = np.sqrt(mean_squared_error(y_test_np, y_pred_np))
mae = mean_absolute_error(y_test_np, y_pred_np)
mse = mean_squared_error(y_test_np, y_pred_np)
pearson_corr, _ = pearsonr(y_test_np, y_pred_np)
r2 = r2_score(y_test_np, y_pred_np)

print("Root Mean Squared Error (RMSE):", rmse)
print("Mean Absolute Error (MAE):", mae)
print("Mean Squared Error (MSE):", mse)
print("Pearson Correlation:", pearson_corr)
print("R-squared (R2):", r2)

Root Mean Squared Error (RMSE): 1.330331297938056
Mean Absolute Error (MAE): 0.9968494775014761
Mean Squared Error (MSE): 1.7697813622735528
Pearson Correlation: 0.8828053773345808
R-squared (R2): 0.7761309605634303


**Root Mean Squared Error (RMSE):** <br>
RMSE measures the average deviation of the predicted values from the actual values. It is the square root of the average of the squared differences between predicted and actual values. In this case, the RMSE is approximately 1.33, which means that, on average, the predicted LN_IC50 values deviate from the actual values by around 1.33 units.

**Mean Absolute Error (MAE):** <br>
MAE measures the average absolute deviation of the predicted values from the actual values. It is the average of the absolute differences between predicted and actual values. In this case, the MAE is approximately 0.997, indicating that, on average, the predicted LN_IC50 values deviate from the actual values by around 0.997 units.

**Mean Squared Error (MSE):** <br>
MSE measures the average of the squared differences between predicted and actual values. In this case, the MSE is approximately 1.77, which is the average squared deviation of the predicted values from the actual values.

**Pearson Correlation:** <br>
Pearson correlation coefficient measures the linear correlation between predicted and actual values. It ranges from -1 to 1, where 1 indicates a perfect positive linear relationship, 0 indicates no linear relationship, and -1 indicates a perfect negative linear relationship. Here, the Pearson correlation coefficient is approximately 0.883, indicating a strong positive linear correlation between predicted and actual values.

**R-squared (R2):** <br>
R-squared is a measure of how well the variation in the dependent variable (LN_IC50) is explained by the independent variables (features). It ranges from 0 to 1, where 1 indicates that all the variability in the dependent variable is explained by the independent variables. In this case, R-squared is approximately 0.776, indicating that around 77.6% of the variability in LN_IC50 is explained by the features used in the model.

Overall, these metrics indicate that the model performs reasonably well in predicting LN_IC50 values, with relatively low errors (RMSE and MAE), a strong linear correlation (Pearson correlation), and a moderate level of explained variability (R-squared).